examples/infix_math_parser.py

```diff
@@ -0,0 +1,190 @@
+performance.
+
+Examples of parsing:
+
+The expression "f_1 + f_2 - 1e-3" is parsed into [['f_1', '+', 'f_2', '-', 0.001]]
+
+The expression "Max(Ln(x) + Lb(Abs(y)), Ceil(Sqrt(garlic) * 3), (potato ** 2) / 4, Abs(cosmic) + 10)"
+is parsed into
+[['Max', [[['Ln', ['x']], '+', ['Lb', [['Abs', ['y']]]]], ['Ceil', [[['Sqrt', ['garlic']], '*', 3]]], [['potato', '**', 2], '/', 4], [['Abs', ['cosmic']], '+', 10]]]]
+"""
+
+from operator import or_
+from typing import ClassVar
+
+from pyparsing import (
+    Forward,
+    Group,
+    Literal,
+    ParserElement,
+    Suppress,
+    delimitedList,
+    infixNotation,
+    one_of,
+    opAssoc,
+    pyparsing_common,
+    Regex,
+)
+
+# Enable Packrat for better performance in recursive parsing
+ParserElement.enablePackrat(None)
+
+
+class InfixExpressionParser:
+    """A class for defining an infix notation parsers."""
+
+    # Supported infix binary operators, i.e., '1+1'. The key is the notation of the operator in infix format,
+    # and the value the notation in parsed format.
+    BINARY_OPERATORS: ClassVar[dict] = {
+        "+": "Add",
+        "-": "Subtract",
+        "*": "Multiply",
+        "/": "Divide",
+        "**": "Power",
+    }
+
+    # Supported infix unary operators, i.e., 'Cos(90)'. The key is the notation of the operator in infix format,
+    # and the value the notation in parsed format.
+    UNARY_OPERATORS: ClassVar[dict] = {
+        "Cos": "Cos",
+        "Sin": "Sin",
+        "Tan": "Tan",
+        "Exp": "Exp",
+        "Ln": "Ln",
+        "Lb": "Lb",
+        "Lg": "Lg",
+        "LogOnePlus": "LogOnePlus",
+        "Sqrt": "Sqrt",
+        "Square": "Square",
+        "Abs": "Abs",
+        "Ceil": "Ceil",
+        "Floor": "Floor",
+        "Arccos": "Arccos",
+        "Arccosh": "Arccosh",
+        "Arcsin": "Arcsin",
+        "Arcsinh": "Arcsinh",
+        "Arctan": "Arctan",
+        "Arctanh": "Arctanh",
+        "Cosh": "Cosh",
+        "Sinh": "Sinh",
+        "Tanh": "Tanh",
+        "Rational": "Rational",
+        "-": "Negate",
+    }
+
+    # Supported infix variadic operators (operators that take one or more comma separated arguments),
+    # i.e., 'Max(1,2, Cos(3)). The key is the notation of the operator in infix format,
+    # and the value the notation in parsed format.
+    VARIADIC_OPERATORS: ClassVar[dict] = {"Max": "Max"}
+
+    def __init__(self):
+        """A parser for infix notation, e.g., the human readable way of notating mathematical expressions.
+
+        The parser can parse infix notation stored in a string. For instance,
+        "Cos(2 + f_1) - 7.2 + Max(f_2, -f_3)" is parsed to the list:
+        ['Cos', [[2, '+', 'f_1']]], '-', 7.2, '+', ['Max', ['f_2', ['-', 'f_3']].
+
+        """
+        # Scope limiters
+        lparen = Suppress("(")
+        rparen = Suppress(")")
+
+        # Define keywords (Note that binary operators must be defined manually)
+        symbols_variadic = set(InfixExpressionParser.VARIADIC_OPERATORS.keys())
```

Comment:
Calling keys() is unnecessary here, since iterating over a dict gives an iterator over its keys.